### PR TITLE
Add validated provenance and reference values

### LINKED ISSUE #166: Add `ValidatedReferenceValues` and `ValidatedProvenanceValues`
The `ProvenanceIR` is essentially a dictionary between reference values and values from all possible provenance formats/versions.

However, depending on whether we are using the `ProvenanceIR` for reference values or values from the provenance format we have different invariants:
- For reference values we want at least one reference value for every value we want to verify to be present. The values we want to verify depend on the provenance format, which the caller needs to specify when validating reference values.
- For values from provenances we have at most one value for every value we want to verify. We allow 0 values, so we can map every possible provenance format to the `ProvenanceIR`.

I want to add two type aliases for `ProvenanceIR`: `ValidatedReferenceValues` and `ValidatedProvenanceValues` to signal these invariants.

Our `ProvenanceIRVerifier` currently verifies every _non-empty_ value in `ValidatedReferenceValue`.

If we impose the following conventions on `ValidatedReferenceValues`, we can use the same `ProvenanceIRVerifier` for every provenance format:

When validating the reference values, depending on the provenance format, we can additionally encode
- whether no value needs to be present (by setting the value to []). This allows to share the same reference values among different provenance formats and we can always add more reference values.
- whether any value at all should be present (by adding a default value).

Alternatively, we could  
- maintain a map from the struct fields of `ProvenanceIR` to bool to indicate whether they need to be verified for a given provenance format, but I don't see much benefit there,
- replace the generic `ProvenanceIRVerifier` by verifiers for every provenance format (e.g. `AmberProveanceVerifier`) calling the individual verification functions one by one, e.g., `verifyBinarySHA256Digest`, `verifyHasBuildCmd`. I think this has a lot of repetition.

### PR DESCRIPTION
**Pull request by mariaschett**
Fixes #166 .

### CODE REVIEW

File: internal/verifier/verifier.go [hunk 1]
```diff
@@ -146,6 +146,38 @@ func (verifier *ProvenanceMetadataVerifier) Verify() (VerificationResult, error)
...
+// and 0 values for every field that is to be ignored.
+type ValidatedReferenceValues = ProvenanceIR
+
+// If a new field is added to `ProvenanceIR`, this has to be updated.
+func (want *ProvenanceIR) ValidateReferenceValues(provenanceFormat string) (*ValidatedReferenceValues, error) {
+
+	// BinarySHA256Digest
+	switch provenanceFormat {
+	default:
+		if len(want.BinarySHA256Digests) <= 1 {
```
Comment (rbehjati) [inline internal/verifier/verifier.go:162]: The condition should be `len(want.BinarySHA256Digests) < 1`, right?

File: internal/verifier/verifier.go [hunk 2]
```diff
@@ -146,6 +146,38 @@ func (verifier *ProvenanceMetadataVerifier) Verify() (VerificationResult, error)
...
 
+// ValidatedProvenanceValues contain exactly one or no value in every field.
+type ValidatedProvenanceValues = ProvenanceIR
+
+// ValidatedReferenceValues contain at least 1 value for every field the provenanceFormat mandates to check,
+// and 0 values for every field that is to be ignored.
+type ValidatedReferenceValues = ProvenanceIR
+
+// If a new field is added to `ProvenanceIR`, this has to be updated.
+func (want *ProvenanceIR) ValidateReferenceValues(provenanceFormat string) (*ValidatedReferenceValues, error) {
```
Comment (rbehjati) [inline internal/verifier/verifier.go:157]: Where does `provenanceFormat` come from? 
All provenance formats are slsa provenances, with different buildTypes. I suggest we use the buildType as the format, and have a reference value for the supported buildTypes as well. WDYT @mariaschett?

Comment (mariaschett) [reply internal/verifier/verifier.go:157]: > Where does provenanceFormat come from?

Options: 
1. We infer the provenanceFormat from the given provenance.
2. The caller decides, which implicitly happens currently, when calling `ProvenanceMetadataVerifier` (we now that this is `SLSAv02`).

> buildType as format / reference value for the buildTypes

Let's discuss this offline.

**Comment by mariaschett** [comment]
This is redudant because of #194 and #165 .

**Pull request closed without merging**